### PR TITLE
APEXCORE-107 Support for adding modules in the DAG through property and json file

### PR DESCRIPTION
**Pull request by tushargosavi**
Support for adding modules in the DAG property file and json file.
Also changed module API to make module as an operator.

### CODE REVIEW

**Comment by tushargosavi** [comment]
@PramodSSImmaneni @sandeepdeshmukh can you review this pull request?

File: api/src/main/java/com/datatorrent/api/Module.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
  * @since 3.3.0
  */
 @InterfaceStability.Evolving
-public interface Module
+public interface Module extends Operator
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/Module.java]: Operator contains execution life cycle methods, which are not appropriate for "Module". Please consider a different design.

**Comment by tweise** [comment]
Looks like this already came up during previous discussion and there wasn't any conclusion. Need to go back to that.

**Comment by pramodin** [comment]
In the current form, where module gets materialized in the logical plan and does not have a runtime aspect to itself directly, it would be difficult to justify/attribute the lifecycle methods to the module. Maybe module and operator can inherit from a common element which could be the building block of DAG. Thats a bigger change and requires more discussions with wider group.

File: api/src/main/java/com/datatorrent/api/Module.java (hunk 1 above)
Comment (tushargosavi) [inline api/src/main/java/com/datatorrent/api/Module.java]: The difference design could be having a common interface, which is shared by all type of Nodes in the DAG. Have updated the pull request with this approach.

File: api/src/main/java/com/datatorrent/api/DagNode.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.api;
+
+/** Marker interface for a node in logical DAG */
+public interface DagNode
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DagNode.java]: Since ports are defined under Operator, this should probably live there also.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java [hunk 3]
```diff
@@ -2187,6 +2190,27 @@ public void populateDAG(LogicalPlan dag)
 
   }
 
+  private DagNode addNode(LogicalPlan dag, String name, DagNode nd)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java]: "nd" ?

File: engine/src/test/java/com/datatorrent/stram/plan/logical/module/TestModuleExpansion.java [hunk 4]
```diff
@@ -184,7 +190,7 @@ public void setPortMemory(int portMemory)
     }
   }
 
-  static class Level2ModuleA implements Module
+  public static class Level2ModuleA extends BaseOperator implements Module
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/plan/logical/module/TestModuleExpansion.java]: Why these changes?

File: api/src/main/java/com/datatorrent/api/DagNode.java (hunk 2 above)
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DagNode.java]: How about naming it Vertex? Other suggestions?

Comment (pramodin) [inline api/src/main/java/com/datatorrent/api/DagNode.java]: I like vertex. How about stream extending edge :).

Comment (sandeshh) [inline api/src/main/java/com/datatorrent/api/DagNode.java]: dag is no more a dag. Vertex +1

File: engine/src/test/java/com/datatorrent/stram/plan/logical/module/TestModuleExpansion.java (hunk 4 above)
Comment (tushargosavi) [inline engine/src/test/java/com/datatorrent/stram/plan/logical/module/TestModuleExpansion.java]: These changes are required for test cases to pass. without public newInstance invocation fails in addOperator.  There are similar test operator which are defined as public for example TestGeneratorInputOperator

I am removing extends BaseOperator from modules, they were leftovers from previous implementation where module was extending operator.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java (hunk 3 above)
Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java]: changed it to vertex.

File: api/src/main/java/com/datatorrent/api/DagNode.java (hunk 2 above)
Comment (tushargosavi) [inline api/src/main/java/com/datatorrent/api/DagNode.java]: Renaming DagNode to Vertex. If I put Vertex in Operator then I can not extend Operator from it, compiler throws cyclic inheritance error.

**Comment by tweise** [comment]
The common interface that we introduce for a vertex in the logical DAG, it could probably need some broader review and discussion. As for the naming, other candidates I have: LogicalDagNode, LogicalOperator.

I prefer to keep the user facing term "operator" stable and consistent. The new interface we discuss is more of an internal detail to process operators in a uniform manner. As such, it would be nice to not even see it as top level interface.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 5]
```diff
@@ -239,7 +239,7 @@ public void sendMetrics(Collection<String> metricNames)
   {
     @SuppressWarnings("FieldNameHidesFieldInSuperclass")
     private static final long serialVersionUID = 1L;
-    private OperatorMeta operatorMeta;
+    private DagNodeMeta operatorMeta;
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: keep OperatorMeta

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 6]
```diff
@@ -778,32 +799,162 @@ public void resetStreamPersistanceOnSinkRemoval(InputPortMeta sinkBeingRemoved)
...
+      return this.outputStreams;
+    }
+
+  }
+
   /**
    * Operator meta object.
    */
-  public final class OperatorMeta implements DAG.OperatorMeta, Serializable
+  public class OperatorMeta extends DagNodeMeta implements DAG.OperatorMeta, Serializable
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: Not clear why we need this distinction and why some of the fields are present here and not in "DagNodeMeta".

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 7]
```diff
@@ -778,32 +799,162 @@ public void resetStreamPersistanceOnSinkRemoval(InputPortMeta sinkBeingRemoved)
     }
   }
 
+  public class DagNodeMeta implements Serializable
+  {
+    protected final LinkedHashMap<InputPortMeta, StreamMeta> inputStreams = new LinkedHashMap<>();
+    protected final LinkedHashMap<OutputPortMeta, StreamMeta> outputStreams = new LinkedHashMap<>();
+    @NotNull
+    protected final String name;
+    protected transient Vertex node;
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: node -> operator

Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: done

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java (hunk 5 above)
Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: The port can belong to Operator or Module. hence used a common meta object in the portmeta. This common meta object contains only
- ports 
- name
- object

The rest of Operator specific fields are in OperatorMeta and Module specific fields are in ModuleMeta.

**Comment by tushargosavi** [comment]
Putting Vertex inside Component , to avoid it being as top level interface. As all node component of the graph inherits from Vertex.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java [hunk 8]
```diff
@@ -2187,6 +2190,27 @@ public void populateDAG(LogicalPlan dag)
 
   }
 
+  private Vertex addNode(LogicalPlan dag, String name, Vertex vertex)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java]: addOperator

Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java]: addVertex?

File: api/src/main/java/com/datatorrent/api/Component.java [hunk 9]
```diff
@@ -116,4 +116,13 @@ public boolean equals(Object obj)
 
   }
 
+  /**
+   * Marker interface for the Node in the DAG. Any object which can be added as a Node in the DAG
+   * needs to implement this interface.
+   */
+  interface Vertex
```
Comment (pramodin) [inline api/src/main/java/com/datatorrent/api/Component.java]: How about moving it into DAG as this is specific to DAG. Component is more general and is implemented by artifacts that are not part of DAG such as security components.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java (hunk 8 above)
Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java]: Changed it to addOperator.

File: api/src/main/java/com/datatorrent/api/Component.java (hunk 9 above)
Comment (tushargosavi) [inline api/src/main/java/com/datatorrent/api/Component.java]: moved the Vertex interface to DAG.

**Comment by davidyan74** [comment]
I think in the field of streaming technology, all of these different nodes in the DAG share the same concept of an operator. So, instead of introducing a new term "vertex" to the platform, how about renaming this interface to "GenericOperator"?

**Comment by pramodin** [comment]
I also support changing the name to GenericOperator

**Comment by tushargosavi** [comment]
changed the name of Vertex to GenericOperator

**Comment by tushargosavi** [comment]
Merged GenericOperatorMeta with OperatorMeta.  ModuleMeta extends from OpeartorMeta. We can merge ModuleMeta and keep a single OperatorMeta but that will involve some changes in more number of files. Let me know if merging ModuleMeta with OperatorMeta if way to go forward.
